coding.md

```diff
@@ -0,0 +1,109 @@
+  variable names. Avoid any variable name that could be confused with an
+  object method (like endswith, startswith, split, upper, index, count).
+
+## Tasks
+* When possible, use full YAML syntax (`key: value`) for task arguments,
+  instead of `key=value` pairs. Some task inputs will require full yaml,
+  it is better to be consistent across the playbooks.
+
+* When writing a task that notifies handlers, the notify statement should be
+  last. Any tags above the notify. Any loop statement should come just before
+  the tags, and any conditional should go above that.
+
+* Task names must not use variables in them.
+
+* All Tasks, Plays, and Handlers must have a name.
+
+* Do not use `ignore_errors`, instead use `failed_when: false`. This prevents
+  red output from being shown on screen, which makes debugging difficult.
+
+* Tasks which just gather data and do not cause change should not register
+  "change" from ansible point of view. Make use of `changed_when: False`.
+
+* No tasks should prompt the user for interaction. The intention of Ursula
+  is to run completely non-interactive.
+
+* Tag expression should be either a single word `tags: thistag`, or a YAML
+  list of tags to apply with each tag on a new line. Do not use `tags=foo`,
+  particularly when including a task file.
+
+* Do not rely on a when conditional to identify that a loop variable is
+  undefined to avoid the task. Use a filter to default the variable to an
+  empty list: `with_items: "{{ foo|default([]) }}"`
+
+## Roles
+* Role dependencies have a cost. Be wary of introducing a new role
+  dependency, particularly if it could be ran from many plays, or if the
+  dependency role includes a large number of tasks.
+
+* Task file includes from a role's `tasks/main.yml` should be avoided
+  unless there is a desire to apply a tag or a conditional to an entire
+  set of logically grouped tasks.
+
+## Modules
+* Prefer the command module over the shell module.
+
+* Files are put in place via the template module, rather than copy, so that
+  all the input files that go to a system are gathered in the same directory
+  (`templates/`)
+
+* Directory structure within templates/ must match (as close as possible) the
+  destination path of the file in question. e.g. if the destination is
+  `/etc/nova/nova.conf`, the file would live in `templates/etc/nova/nova.conf`
+  (note the lack of `.j2` suffix)
+
+## Testing
+* When adding new features to Ursula, it is desireable that they are
+  feature flagged in a way to prevent immediate implementation of the
+  feature in existing installs.
+
+* That said, when at all possible it is desireable for our CI runs to
+  at least exercise the Ansible code used to deliver a new feature, even if
+  the feature itself is not tested in CI.
+
+* Be cognizant of CI runtimes. The more tests, the longer the run, the slower
+  the changes will flow. Try to balance good coverage with quick tests for
+  CI runs.
```

Comment:
Even with more CI resources, a lot of tests means a longer return. It's a fine balance to strike. What I'm hoping to see is more nightly tests of a tree that are able to do deeper integrations and slower tests.